### PR TITLE
Check compatibility

### PR DESCRIPTION
**Pull request by harjinder-hari**
Added a feature to generate license filter information along with regular license analysis.

Now, the stack license analysis end point will also accept recommended alternate and companion packages in the input. If the stack license is possible ( i.e. no unknown and no conflict ) then the service will also generate information for license based filters.

The license filter info will have three pieces:
 1. unknown licenses packages
 2. conflict packages
 3. compatible packages

Thanks,
Harjinder

### CODE REVIEW

File: src/rest_api.py [hunk 1]
```diff
@@ -20,6 +20,13 @@
 app.config.from_object('config')
 CORS(app)
 
+global stack_license_analyzer
```
Comment (miteshvp) [inline src/rest_api.py]: where is this global variable used?

**Comment by miteshvp** [comment]
[test]

File: src/license_analysis.py [hunk 2]
```diff
@@ -526,3 +526,75 @@ def compute_representative_license(self, input_licenses):
         output['representative_license'] = None
         return output
 
+    def _get_compatibility_classes(self, input_license):
+        list_comp_classes = []
+        for item in self.dict_compatibility_classes.items():
+            if input_license in item[1]:
+                list_comp_classes.append(item[0])
```
Comment (jpopelka) [inline src/license_analysis.py]: Something like:

```python
for license, list_compatibles in self.dict_compatibility_classes.items():
    if input_license in list_compatibles:
        list_comp_classes.append(license)
```

would make the code more understandable for somebody who sees it for the first time.

You could also use the list comprehension to squash the 4 lines into one, but that's optional.

`list_comp_classes = [license for license, list_compatibles in self.dict_compatibility_classes.items() if input_license in list_compatibles]`

File: src/license_analysis.py [hunk 3]
```diff
@@ -526,3 +526,75 @@ def compute_representative_license(self, input_licenses):
...
+            'status': 'Failure',
+            'reason': 'Input is invalid',
+            'unknown_licenses': [],
+            'conflict_licenses': [],
+            'compatible_licenses': [],
+            'synonyms': []
+        }
+        if lic_a is None:
+            return output
+        if len(list_lic_b) == 0:
```
Comment (jpopelka) [inline src/license_analysis.py]: You can check whether a list variable is not `None` and not empty in one go with: `if not list_variable:`
Also you can test both `lic_a` and `list_lic_b` in one condition with:

```python
if not lic_a or not list_lic_b:
    return output
```

File: src/license_analysis.py [hunk 4]
```diff
@@ -526,3 +526,75 @@ def compute_representative_license(self, input_licenses):
...
+            return output
+
+        # find synonyms
+        lic_a = self.find_synonym(lic_a)
+        list_lic_b_sysnonyms = [self.find_synonym(y) for y in list_lic_b]
```
Comment (jpopelka) [inline src/license_analysis.py]: There's a typo in `list_lic_b_sysnonyms` (should be `list_lic_b_synonyms`)

File: src/license_analysis.py [hunk 5]
```diff
@@ -526,3 +526,75 @@ def compute_representative_license(self, input_licenses):
...
+
+        # we will now work with the synonyms
+        list_lic_b = list_lic_b_sysnonyms
+
+        # now, we need to find compatibility class for lic_b and every lic_b
+        lic_a_group = self._get_compatibility_classes(lic_a)
+        assert(len(lic_a_group) > 0)
+        lic_b_groups = {}
+        for lic_b in list_lic_b:
+            lic_b_groups[lic_b] = self._get_compatibility_classes(lic_b)
```
Comment (jpopelka) [inline src/license_analysis.py]: You can use [Dict Comprehensions](https://www.python.org/dev/peps/pep-0274/)

```python
lic_b_groups = {lic_b: self._get_compatibility_classes(lic_b) for lic_b in list_lic_b}
```

File: src/license_analysis.py [hunk 6]
```diff
@@ -526,3 +526,75 @@ def compute_representative_license(self, input_licenses):
...
+
+        # initialize dict that maps every lic_b to one of lic_a's compatibility classes
+        list_items = [(x, []) for x in lic_a_group]
+        map_compatibility = dict(list_items)
```
Comment (jpopelka) [inline src/license_analysis.py]: [Dict Comprehensions](https://www.python.org/dev/peps/pep-0274/):

`map_compatibility = {x: [] for x in lic_a_group}`

File: src/stack_license.py [hunk 7]
```diff
@@ -61,69 +37,166 @@ def compute_stack_license(payload):
...
-            if la_output['representative_license'] is None:
-                # This will later indicate that no need to compute stack license
-                is_stack_license_possible = False
+                list_conflict_lic = la_output['conflict_licenses']
+                list_conflict_pkg = []
+                for lic_tuple in list_conflict_lic:
+                    pkg_group = {
+                        dict_lic_pkg[lic_tuple[0]]: lic_tuple[0],
+                        dict_lic_pkg[lic_tuple[1]]: lic_tuple[1]
+                    }
```
Comment (jpopelka) [inline src/stack_license.py]: Rather:

```python
             for a, b in list_conflict_lic:
                 pkg_group = {
                     dict_lic_pkg[a]: a,
                     dict_lic_pkg[b]: b
                 }
```

(replace `a` and `b` with some meaningful names)

File: src/stack_license.py [hunk 8]
```diff
@@ -61,69 +37,166 @@ def compute_stack_license(payload):
...
+                    list_conflict_pkg.append(pkg_group)
+                output['conflict_packages'] = list_conflict_pkg
+                output['status'] = 'StackLicenseConflict'
+
+            if (len(la_output['outlier_licenses'])) > 0:
+                outlier_pkg = {}
+                for lic in la_output['outlier_licenses']:
+                    outlier_pkg[dict_lic_pkg[lic]] = lic
```
Comment (jpopelka) [inline src/stack_license.py]: `outlier_pkg = {dict_lic_pkg[lic]: lic for lic in la_output['outlier_licenses']}`

File: src/rest_api.py (hunk 1 above)
Comment (harjinder-hari) [reply src/rest_api.py]: Good point :) I have removed this variable.

File: src/license_analysis.py (hunk 3 above)
Comment (harjinder-hari) [reply src/license_analysis.py]: Done.

File: src/license_analysis.py (hunk 4 above)
Comment (harjinder-hari) [reply src/license_analysis.py]: Done.

File: src/license_analysis.py (hunk 5 above)
Comment (harjinder-hari) [reply src/license_analysis.py]: Done.

File: src/license_analysis.py (hunk 6 above)
Comment (harjinder-hari) [reply src/license_analysis.py]: Done.

File: src/stack_license.py (hunk 7 above)
Comment (harjinder-hari) [reply src/stack_license.py]: Done.

File: src/stack_license.py (hunk 8 above)
Comment (harjinder-hari) [reply src/stack_license.py]: I need to change dict_lic_pkg from lic->pkg to lic->list_pkgs. So, could not use the dict comprehension here.

**Comment by harjinder-hari** [comment]
Thanks @miteshvp @jpopelka ! I have addressed the review comments. Can you please take a look at the latest commit ?

Thanks,
Harjinder

File: src/stack_license.py [hunk 9]
```diff
@@ -74,6 +74,16 @@ def _check_compatibility(self, stack_license, other_packages):
...
         return output
 
+    @staticmethod
+    def _generate_conflict_pairs(list_pkg1, list_pkg2):
+        list_tuples = []
+        for pkg1 in list_pkg1:
+            for pkg2 in list_pkg2:
+                list_tuples.append((pkg1, pkg2))
+
+        return
```
[the remark this answers is not in the document]
Comment (harjinder-hari) [reply src/stack_license.py]: Thanks Jiri ! That function actually is not used by any one. Sorry for the confusion. I have remove that function.

If things looks good then can you please merge this PR ?

Thanks,
Harjinder

**Pull request merged**